src/draco/io/stl_encoder.cc

```diff
@@ -0,0 +1,125 @@
+// Copyright 2016 The Draco Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "draco/io/stl_encoder.h"
+
+#include <memory>
+#include <sstream>
+
+#include "draco/io/file_writer_factory.h"
+#include "draco/io/file_writer_interface.h"
+
+namespace draco {
+
+StlEncoder::StlEncoder()
+    : out_buffer_(nullptr), in_point_cloud_(nullptr), in_mesh_(nullptr) {}
+
+bool StlEncoder::EncodeToFile(const Mesh &mesh, const std::string &file_name) {
+  in_mesh_ = &mesh;
+  std::unique_ptr<FileWriterInterface> file =
+      FileWriterFactory::OpenWriter(file_name);
+  if (!file) {
+    return false;  // File couldn't be opened.
+  }
+  // Encode the mesh into a buffer.
+  EncoderBuffer buffer;
+  if (!EncodeToBuffer(mesh, &buffer)) {
+    return false;
+  }
+  // Write the buffer into the file.
+  file->Write(buffer.data(), buffer.size());
+  return true;
+}
+
+bool StlEncoder::EncodeToBuffer(const Mesh &mesh, EncoderBuffer *out_buffer) {
+  in_mesh_ = &mesh;
+  out_buffer_ = out_buffer;
+  if (!EncodeInternal()) {
+    return ExitAndCleanup(false);
+  }
+  return ExitAndCleanup(true);
+}
+bool StlEncoder::EncodeInternal() {
+  // Write STL header.
+  std::stringstream out;
+  out << "generated using draco_decoder           "; // 40 bytes +
+  out << "                                        "; // 40 bytes == header
+  const std::string header_str = out.str();
+  buffer()->Encode(header_str.data(), header_str.length());
```

Comment:
can we just hardcode the encoded length to 80 to make sure the correct number of bytes is written?

Reply:
I will check it out!

Reply:
New approach: out << std::left << std::setw(80) << "generated using Draco";